File: typescript/rna-transcription/rna-transcription.ts
type DnaNucleotide = "G" | "C" | "T" | "A"
type RnaNucleotide = "C" | "G" | "A" | "U"
const NucleotidesMap: Record<DnaNucleotide, RnaNucleotide> = {
    "G": "C",
    "C": "G",
    "T": "A",
    "A": "U"
}

class Transcriptor {
    toRna = (dna: string): string => {
        return dna.replace(/./g, this.convertNucleotide)
    }

    convertNucleotide = (strand: string): RnaNucleotide => {
        if (this.isNucleotidesOfDna(strand)) {
            return NucleotidesMap[strand]
        } else {
            throw new Error("Invalid input DNA.")
        }
    }

    isNucleotidesOfDna = (strand: string): strand is DnaNucleotide => {
        return strand in NucleotidesMap
    }
}

export default Transcriptor

File: go/rna-transcription/rna_transcription.go
package strand

var toRnaMap = map[rune]rune{
	'G': 'C',
	'C': 'G',
	'T': 'A',
	'A': 'U',
}

// ToRNA convert input dna string and return a rna string
func ToRNA(dna string) string {
	rna := make([]rune, len(dna))
	for k, r := range dna {
		rna[k] = toRnaMap[r]
	}
	return string(rna)

}
